helix-core/src/syntax/span.rs

```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
     }
 }
 
+impl SpanIter {
+    fn start_span(&mut self, span: Span) {
+        use HighlightEvent::*;
+
+        debug_assert!(span.start <= span.end);
+        self.event_queue
+            .push_back(HighlightStart(Highlight(span.scope)));
+        self.range_ends.push(span.end);
+    }
+
+    fn process_range_end(&mut self, end: usize) -> HighlightEvent {
+        use HighlightEvent::*;
+
+        let start = replace(&mut self.cursor, end);
+        if start != end {
+            debug_assert!(start < end);
+            self.event_queue.push_back(HighlightEnd);
+            Source { start, end }
+        } else {
+            HighlightEnd
+        }
+    }
+
+    // Any subslices that end before intersect span needs to be subsliced, consume the
+    // left part of the subslice and leave the right.
+    fn partition_spans_at(&mut self, intersect: usize) {
+        let first_partioned_span = self.spans[self.index];
+
+        let mut i = self.index;
+        while let Some(span) = self.spans.get_mut(i) {
+            if span.start != self.cursor || span.end < intersect {
+                break;
+            }
+
+            let mut partinoed_span = *span;
+            partinoed_span.end = intersect;
+            span.start = intersect;
+
+            self.start_span(partinoed_span);
+            i += 1;
+        }
+
+        let subslices = i - self.index;
+
+        // When spans are subsliced, the span Vec may need to be re-sorted
+        // because the `range.start` may now be greater than some `range.start`
+        // later in the Vec. This is not a classic "sort": we take several
+        // shortcuts to improve the runtime so that the sort may be done in
+        // time linear to the cardinality of the span Vec. Practically speaking
+        // the runtime is even better since we only scan from `self.index` to
+        // the first element of the Vec with a `range.start` after this range.
+        let mut after = None;
+        let intersect_pos = Span {
```

Comment:
I think this would read clearer as `intersect_span`

Reply:
fixed in 20302313b5d274d231c8c2e505a5306601c0173d.

This made me notice that the term `range` is kind of used randomly instead of `span`. I think it might be nice to consistently use the term `span` instead. I did that in 7de6f2a8ab7f47342d5ab017fca87cf05b9a7f5b and also expanded on the documentation of `partition_spans_at`. Do you think the renaming is better or should I revert that commit (except for the documentation)?